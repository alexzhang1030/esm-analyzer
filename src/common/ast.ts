import { walk } from 'estree-walker'
import type { ASTNode, ASTNodeLocation, WalkCallback, t } from '@/types'

export function walkAST(node: ASTNode, handlers: {
  enter?: WalkCallback
  leave?: WalkCallback
}): ASTNode {
  return walk(node, handlers)
}

export function getASTNodeLocation(node: ASTNode): ASTNodeLocation {
  const { start: _s, end: _e } = node.loc!
  return {
    start: {
      line: _s.line,
      column: _s.column,
      index: node.start!,
    },
    end: {
      line: _e.line,
      column: _e.column,
      index: node.end!,
    },
  }
}

// ast is
export function isImportDeclaration(node: ASTNode): node is t.ImportDeclaration {
  return node.type === 'ImportDeclaration'
}

export function isExportNamedDeclaration(node: ASTNode): node is t.ExportNamedDeclaration {
  return node.type === 'ExportNamedDeclaration'
}

export function isExportDefaultDeclaration(node: ASTNode): node is t.ExportDefaultDeclaration {
  return node.type === 'ExportDefaultDeclaration'
}

export function isExportAllDeclaration(node: ASTNode): node is t.ExportAllDeclaration {
  return node.type === 'ExportAllDeclaration'
}

export function isVariableDeclaration(node: ASTNode): node is t.VariableDeclaration {
  return node.type === 'VariableDeclaration'
}

export function isIdentifier(node: ASTNode): node is t.Identifier {
  return node.type === 'Identifier'
}
