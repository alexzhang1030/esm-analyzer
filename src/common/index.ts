export * from './ast'
export * from './lang'
export * from './parse'
