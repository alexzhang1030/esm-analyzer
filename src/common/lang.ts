import type { AcceptableLang } from '..'

const REGEX_LANG_TS = /.?[mc]?tsx?$/
const REGEX_LANG_JSX = /.?[jt]sx$/
const REGEX_ACCEPTABLE_LANG = /.?[mc]?[jt]sx?$/

export const isTS = (lang: string) => REGEX_LANG_TS.test(lang)
export const isJSX = (lang: string) => REGEX_LANG_JSX.test(lang)
export const isAcceptableLang = (filename: string) => REGEX_ACCEPTABLE_LANG.test(filename)

export function getLangByFileName(filename: string) {
  return filename.split('.').pop() as AcceptableLang
}
