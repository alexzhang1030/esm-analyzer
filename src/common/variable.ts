import type { Primitive } from 'type-fest'
import { isIdentifier } from './ast'
import type { ASTNode } from '@/types'

type PrimitiveValue =
  | Primitive
  | null | undefined

export interface PrimitiveVariableValue {
  type: 'StringLiteral' | 'NumericLiteral' | 'BooleanLiteral' | 'NullLiteral' | 'ObjectExpression' | 'ArrayExpression'
  value: PrimitiveValue | ResolveVariableDeclaration
}

export interface ObjectExpressionVariableValue {
  type: 'ObjectExpression'
  value: {
    [key: string]: ResolveVariableDeclaration
  }
}

export interface ArrayExpressionVariableValue {
  type: 'ArrayExpression'
  value: ResolveVariableDeclaration[]
}

export interface CallExpressionVariableValue {
  type: 'CallExpression'
  callee: string
  arguments: ResolveVariableDeclaration[]
}

export type ResolveVariableDeclaration =
  | PrimitiveVariableValue
  | CallExpressionVariableValue
  | ObjectExpressionVariableValue
  | ArrayExpressionVariableValue
  | null

export function resolveVariableDeclarationValue(node?: ASTNode | null): ResolveVariableDeclaration {
  if (!node)
    return null
  switch (node.type) {
    case 'StringLiteral':
    case 'NumericLiteral':
    case 'BooleanLiteral':
      return {
        type: node.type,
        value: node.value,
      }
    case 'NullLiteral':
      return {
        type: node.type,
        value: null,
      }
    case 'ObjectExpression':
      return {
        type: node.type,
        value: node.properties.reduce((result, property) => {
          if (!property || property.type !== 'ObjectProperty')
            return result
          if (!isIdentifier(property.key))
            return result
          if (!property.value)
            return result
          result[property.key.name] = resolveVariableDeclarationValue(property.value)
          return result
        }, {} as ObjectExpressionVariableValue['value']),
      }
    case 'ArrayExpression':
      return {
        type: node.type,
        value: node.elements.map((item) => {
          return resolveVariableDeclarationValue(item)
        }),
      }
    case 'CallExpression':
      if (!node.callee || !isIdentifier(node.callee))
        return null
      return {
        type: node.type,
        callee: node.callee.name,
        arguments: node.arguments.map((item) => {
          return resolveVariableDeclarationValue(item)
        }),
      }
  }
  return null
}
