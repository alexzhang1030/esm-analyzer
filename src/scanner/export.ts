import type { ResolveVariableDeclaration } from '@/common'
import { getASTNodeLocation, isExportAllDeclaration, isExportDefaultDeclaration, isExportNamedDeclaration, isIdentifier, resolveVariableDeclarationValue } from '@/common'
import type { ASTNode, ASTNodeLocation, t } from '@/types'
import { loop } from '@/utils'

export type ScanExportType =
  | 'ExportNamedDeclaration'
  | 'ExportAllDeclaration'
  | 'ExportDefaultDeclaration'

export interface ScanExportNamedResult {
  type: 'ExportNamedDeclaration'
  kind: t.VariableDeclaration['kind']
  declarations: {
    name: string
    init: ResolveVariableDeclaration
  }[]
}

export interface ScanExportAllResult {
  type: 'ExportAllDeclaration'
  source: string
}

export interface ScanExportDefaultIdentifierResult {
  type: 'ExportDefaultDeclaration'
  subType: 'Identifier'
  id: string
}

export interface ScanExportDefaultObjectResult {
  type: 'ExportDefaultDeclaration'
  subType: 'ObjectExpression'
  properties: {
    key: string
    value: ResolveVariableDeclaration
  }[]
}

export type ScanExportResult = (
  | ScanExportNamedResult
  | ScanExportAllResult
  | ScanExportDefaultIdentifierResult
  | ScanExportDefaultObjectResult
) & {
  loc: ASTNodeLocation
}

export interface ScanExportConfig {
  includeType?: ScanExportType[]
  excludeType?: ScanExportType[]
}

export function scanExport(node: ASTNode, config?: ScanExportConfig): ScanExportResult | null {
  if (
    !isExportNamedDeclaration(node)
    && !isExportDefaultDeclaration(node)
    && !isExportAllDeclaration(node)
  )
    return null
  if (config?.includeType && !config.includeType.includes(node.type))
    return null
  if (config?.excludeType && config.excludeType.includes(node.type))
    return null
  if (node.type === 'ExportNamedDeclaration') {
    if (!node.declaration || node.declaration.type !== 'VariableDeclaration')
      return null
    const vars: ScanExportNamedResult['declarations'] = []
    loop(node.declaration.declarations, (declaration) => {
      if (!isIdentifier(declaration.id))
        return
      const result = {
        name: declaration.id.name,
        init: resolveVariableDeclarationValue(declaration.init),
      }
      if (!result.init)
        return
      vars.push(result)
    })
    return {
      type: 'ExportNamedDeclaration',
      declarations: vars,
      kind: node.declaration.kind,
      loc: getASTNodeLocation(node),
    }
  }
  else if (node.type === 'ExportAllDeclaration') {
    return {
      type: 'ExportAllDeclaration',
      source: node.source.value,
      loc: getASTNodeLocation(node),
    }
  }
  else {
    if (node.declaration.type === 'Identifier') {
      return {
        type: 'ExportDefaultDeclaration',
        subType: 'Identifier',
        id: node.declaration.name,
        loc: getASTNodeLocation(node),
      }
    }
    else if (node.declaration.type === 'ObjectExpression') {
      return {
        type: 'ExportDefaultDeclaration',
        subType: 'ObjectExpression',
        loc: getASTNodeLocation(node),
        properties: node.declaration.properties.map((property) => {
          if (property.type !== 'ObjectProperty' || !isIdentifier(property.key))
            return null
          return {
            key: property.key.name,
            value: resolveVariableDeclarationValue(property.value),
          }
        }).filter(Boolean) as ScanExportDefaultObjectResult['properties'],
      }
    }
  }
  return null
}
