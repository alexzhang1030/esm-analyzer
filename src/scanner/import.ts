import { getASTNodeLocation, isImportDeclaration } from '@/common'
import type { ASTNode, ASTNodeLocation, t } from '@/types'

import { loop } from '@/utils'

interface ScanResultBase {
  source: string
  loc: ASTNodeLocation
}

export interface ScanImportResultDefault extends ScanResultBase {
  type: 'default'
  local: string // local name
}

export interface ScanImportResultNamespace extends ScanResultBase {
  type: 'namespace'
  local: string // local name
}

export interface ScanImportResultImport extends ScanResultBase {
  type: 'import'
  subType: 'id' | 'string'
  isType: boolean
  local: string // local name
  imported: string // imported name
}

export type ScanImportResultItem = ScanImportResultDefault | ScanImportResultNamespace | ScanImportResultImport

// import bar from 'foo'
function resolveDefaultSpecifier(node: t.ImportDefaultSpecifier) {
  return node.local.name
}

// import * as bar from 'foo'
function resolveNamespaceSpecifier(node: t.ImportNamespaceSpecifier) {
  return node.local.name
}

// import { bar } from 'foo'
function resolveImportSpecifier(node: t.ImportSpecifier) {
  let imported: string
  let subType: ScanImportResultImport['subType'] = 'id'
  const isType = node.importKind === 'type'
  if (node.imported.type === 'Identifier') {
    imported = node.imported.name
  }
  else {
    imported = node.imported.value
    subType = 'string'
  }
  return {
    imported,
    local: node.local.name,
    subType,
    isType,
  }
}

export interface ScanImportConfig {
  includeSource?: string[]
  excludeSource?: string[]
  skipType?: boolean
}

export const defaultConfig: Required<ScanImportConfig> = {
  includeSource: [],
  excludeSource: [],
  skipType: false,
}

export function scanImport(node: ASTNode, config: ScanImportConfig = defaultConfig): ScanImportResultItem[] | null {
  if (!isImportDeclaration(node))
    return null
  const items = node.specifiers
  if (!items.length)
    return null
  const result: ScanImportResultItem[] = []
  const source = node.source.value
  const isType = node.importKind === 'type'
  if (config.includeSource?.length && !config.includeSource.includes(source))
    return null
  if (config.excludeSource?.length && config.excludeSource.includes(source))
    return null
  if (config.skipType && isType)
    return null
  loop(items, (item) => {
    if (item.type === 'ImportDefaultSpecifier') {
      result.push({
        type: 'default',
        source,
        local: resolveDefaultSpecifier(item),
        loc: getASTNodeLocation(item),
      })
    }
    else if (item.type === 'ImportNamespaceSpecifier') {
      result.push({
        type: 'namespace',
        source,
        local: resolveNamespaceSpecifier(item),
        loc: getASTNodeLocation(item),
      })
    }
    else if (item.type === 'ImportSpecifier') {
      const { imported, local, subType, isType: subIsType } = resolveImportSpecifier(item)
      if (config.skipType && subIsType)
        return
      result.push({
        type: 'import',
        loc: getASTNodeLocation(item),
        source,
        imported,
        local,
        subType,
        isType: !isType ? subIsType : isType,
      })
    }
  })
  return result
}
