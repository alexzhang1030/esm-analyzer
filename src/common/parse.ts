import { type ParserOptions, type ParserPlugin, parse } from '@babel/parser'
import { isAcceptableLang, isJSX, isTS } from './lang'
import { walkAST } from './ast'
import type { ASTNode } from '@/types'
import { loop } from '@/utils'

export function getBabelParsePlugins(lang: string) {
  const plugins: ParserPlugin[] = []
  if (isTS(lang))
    plugins.push('typescript')
  if (isJSX(lang))
    plugins.push('jsx')
  return plugins
}

export function babelParse(code: string, lang: string) {
  const options: ParserOptions = {
    plugins: getBabelParsePlugins(lang),
    sourceType: 'module',
  }
  return parse(code, options)
}

type NonNullable<T> = T extends null | undefined ? never : T

type LoadScannerReturnType<T extends ((...args: any[]) => any)[]> = NonNullable<ReturnType<T[number]>>[]

export function loadScanners<
  Func extends (node: ASTNode, ...args: any[]) => any | null = (node: ASTNode, ...args: any[]) => any | null,
  Funcs extends Func[] = Func[],
>(code: string, lang: string, scanner: Funcs) {
  if (!isAcceptableLang(lang))
    throw new Error(`[ESM Analyzer] Unsupported language: ${lang}`)
  const result = Array.from({ length: scanner.length }).fill(null).map<any>(() => [] as any)
  const ast = babelParse(code, lang)
  walkAST(ast, {
    enter(node) {
      loop(scanner, (scanner, index) => {
        const res = scanner(node)
        if (res) {
          if (Array.isArray(res))
            result[index].push(...res)
          else
            result[index].push(res)
        }
      })
    },
  })
  return result as LoadScannerReturnType<Funcs>
}

export function loadScanner<
  Func extends (node: ASTNode, ...args: any[]) => any | null = (node: ASTNode, ...args: any[]) => any | null,
  >(code: string, lang: string, scanner: Func) {
  return loadScanners(code, lang, [scanner])[0]
}
