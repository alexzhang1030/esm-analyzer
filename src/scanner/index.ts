import type { ScanImportConfig, ScanImportResult } from './import'
import { scanImport } from './import'
import type { ScanVariableDeclarationConfig, ScanVariableDeclarationResult } from './variable'
import { scanVariableDeclaration } from './variable'
import type { ScanExportConfig } from './export'
import { scanExport } from './export'
import type { AcceptableLang } from '@/types'
import { loadScanners } from '@/common'

interface ScanConfig {
  import?: ScanImportConfig
  variable?: ScanVariableDeclarationConfig
  export?: ScanExportConfig
}

export function scan(code: string, lang: AcceptableLang, config?: ScanConfig) {
  const [imports, variables, exports] = loadScanners(code, lang, [
    node => scanImport(node, config?.import),
    node => scanVariableDeclaration(node, config?.variable),
    node => scanExport(node, config?.export),
  ])
  return {
    imports,
    variables,
    exports,
  } as {
    imports: ScanImportResult[]
    variables: ScanVariableDeclarationResult[]
    exports: ScanVariableDeclarationResult[]
  }
}

export * from './import'
export * from './variable'
