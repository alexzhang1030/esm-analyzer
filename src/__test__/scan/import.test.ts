import { loadScanner } from '@/common'
import { scan, scanImport } from '@/scanner'

describe('scan import', () => {
  test('import default', () => {
    const code = 'import foo from \'bar\''
    expect(scan(code, 'js').imports).toMatchInlineSnapshot(`
      [
        {
          "loc": {
            "end": {
              "column": 10,
              "index": 10,
              "line": 1,
            },
            "start": {
              "column": 7,
              "index": 7,
              "line": 1,
            },
          },
          "local": "foo",
          "source": "bar",
          "type": "default",
        },
      ]
    `)
  })
  test('import namespace', () => {
    const code = 'import * as foo from \'bar\''
    expect(scan(code, 'js').imports[0]).toMatchInlineSnapshot(`
      {
        "loc": {
          "end": {
            "column": 15,
            "index": 15,
            "line": 1,
          },
          "start": {
            "column": 7,
            "index": 7,
            "line": 1,
          },
        },
        "local": "foo",
        "source": "bar",
        "type": "namespace",
      }
    `)
  })
  test('import id', () => {
    const code = `
      import { foo, foo1 as foo2, "foo3" as foo4 } from 'bar'
      import { bar1 as bar2 } from 'bar'
      import { "baz1" as baz2 } from 'bar'
    `
    expect(scan(code, 'js')).toMatchInlineSnapshot(`
      {
        "imports": [
          {
            "imported": "foo",
            "isType": false,
            "loc": {
              "end": {
                "column": 18,
                "index": 19,
                "line": 2,
              },
              "start": {
                "column": 15,
                "index": 16,
                "line": 2,
              },
            },
            "local": "foo",
            "source": "bar",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "foo1",
            "isType": false,
            "loc": {
              "end": {
                "column": 32,
                "index": 33,
                "line": 2,
              },
              "start": {
                "column": 20,
                "index": 21,
                "line": 2,
              },
            },
            "local": "foo2",
            "source": "bar",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "foo3",
            "isType": false,
            "loc": {
              "end": {
                "column": 48,
                "index": 49,
                "line": 2,
              },
              "start": {
                "column": 34,
                "index": 35,
                "line": 2,
              },
            },
            "local": "foo4",
            "source": "bar",
            "subType": "string",
            "type": "import",
          },
          {
            "imported": "bar1",
            "isType": false,
            "loc": {
              "end": {
                "column": 27,
                "index": 90,
                "line": 3,
              },
              "start": {
                "column": 15,
                "index": 78,
                "line": 3,
              },
            },
            "local": "bar2",
            "source": "bar",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "baz1",
            "isType": false,
            "loc": {
              "end": {
                "column": 29,
                "index": 133,
                "line": 4,
              },
              "start": {
                "column": 15,
                "index": 119,
                "line": 4,
              },
            },
            "local": "baz2",
            "source": "bar",
            "subType": "string",
            "type": "import",
          },
        ],
      }
    `)
  })
  test('combine', () => {
    const code = `
      import foo from 'bar'
      import * as foo2 from 'bar1'
      import { foo3 } from 'bar2'
      import { foo3 as foo4 } from 'bar3'
    `
    expect(scan(code, 'js')).toMatchInlineSnapshot(`
      {
        "imports": [
          {
            "loc": {
              "end": {
                "column": 16,
                "index": 17,
                "line": 2,
              },
              "start": {
                "column": 13,
                "index": 14,
                "line": 2,
              },
            },
            "local": "foo",
            "source": "bar",
            "type": "default",
          },
          {
            "loc": {
              "end": {
                "column": 22,
                "index": 51,
                "line": 3,
              },
              "start": {
                "column": 13,
                "index": 42,
                "line": 3,
              },
            },
            "local": "foo2",
            "source": "bar1",
            "type": "namespace",
          },
          {
            "imported": "foo3",
            "isType": false,
            "loc": {
              "end": {
                "column": 19,
                "index": 83,
                "line": 4,
              },
              "start": {
                "column": 15,
                "index": 79,
                "line": 4,
              },
            },
            "local": "foo3",
            "source": "bar2",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "foo3",
            "isType": false,
            "loc": {
              "end": {
                "column": 27,
                "index": 125,
                "line": 5,
              },
              "start": {
                "column": 15,
                "index": 113,
                "line": 5,
              },
            },
            "local": "foo4",
            "source": "bar3",
            "subType": "id",
            "type": "import",
          },
        ],
      }
    `)
  })
  test('import type', () => {
    const code = `
      import type { foo1, foo2 } from 'foo'
      import { type foo3, foo4_real }  from 'bar'
    `
    expect(scan(code, 'ts')).toMatchInlineSnapshot(`
      {
        "imports": [
          {
            "imported": "foo1",
            "isType": true,
            "loc": {
              "end": {
                "column": 24,
                "index": 25,
                "line": 2,
              },
              "start": {
                "column": 20,
                "index": 21,
                "line": 2,
              },
            },
            "local": "foo1",
            "source": "foo",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "foo2",
            "isType": true,
            "loc": {
              "end": {
                "column": 30,
                "index": 31,
                "line": 2,
              },
              "start": {
                "column": 26,
                "index": 27,
                "line": 2,
              },
            },
            "local": "foo2",
            "source": "foo",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "foo3",
            "isType": true,
            "loc": {
              "end": {
                "column": 24,
                "index": 69,
                "line": 3,
              },
              "start": {
                "column": 15,
                "index": 60,
                "line": 3,
              },
            },
            "local": "foo3",
            "source": "bar",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "foo4_real",
            "isType": false,
            "loc": {
              "end": {
                "column": 35,
                "index": 80,
                "line": 3,
              },
              "start": {
                "column": 26,
                "index": 71,
                "line": 3,
              },
            },
            "local": "foo4_real",
            "source": "bar",
            "subType": "id",
            "type": "import",
          },
        ],
      }
    `)
  })
})

describe('pass config', () => {
  test('includeSource', () => {
    const code = `
      import { a } from 'vue'
      import { b } from 'react'
      import { c } from '@vueuse/core'
    `
    const result = loadScanner(code, 'js', [node => scanImport(node, {
      includeSource: ['vue', '@vueuse/core'],
    })])
    expect(result).toMatchInlineSnapshot(`
      [
        [
          {
            "imported": "a",
            "isType": false,
            "loc": {
              "end": {
                "column": 16,
                "index": 17,
                "line": 2,
              },
              "start": {
                "column": 15,
                "index": 16,
                "line": 2,
              },
            },
            "local": "a",
            "source": "vue",
            "subType": "id",
            "type": "import",
          },
          {
            "imported": "c",
            "isType": false,
            "loc": {
              "end": {
                "column": 16,
                "index": 79,
                "line": 4,
              },
              "start": {
                "column": 15,
                "index": 78,
                "line": 4,
              },
            },
            "local": "c",
            "source": "@vueuse/core",
            "subType": "id",
            "type": "import",
          },
        ],
      ]
    `)
  })
})
