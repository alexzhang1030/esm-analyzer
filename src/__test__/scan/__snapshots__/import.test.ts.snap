// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pass config > excludeSource 1`] = `
[
  [
    {
      "imported": "b",
      "isType": false,
      "loc": {
        "end": {
          "column": 16,
          "index": 47,
          "line": 3,
        },
        "start": {
          "column": 15,
          "index": 46,
          "line": 3,
        },
      },
      "local": "b",
      "source": "react",
      "subType": "id",
      "type": "import",
    },
  ],
]
`;

exports[`pass config > includeSource 1`] = `
[
  {
    "imported": "a",
    "isType": false,
    "loc": {
      "end": {
        "column": 16,
        "index": 17,
        "line": 2,
      },
      "start": {
        "column": 15,
        "index": 16,
        "line": 2,
      },
    },
    "local": "a",
    "source": "vue",
    "subType": "id",
    "type": "import",
  },
  {
    "imported": "c",
    "isType": false,
    "loc": {
      "end": {
        "column": 16,
        "index": 79,
        "line": 4,
      },
      "start": {
        "column": 15,
        "index": 78,
        "line": 4,
      },
    },
    "local": "c",
    "source": "@vueuse/core",
    "subType": "id",
    "type": "import",
  },
  {
    "imported": "d",
    "isType": false,
    "loc": {
      "end": {
        "column": 19,
        "index": 82,
        "line": 4,
      },
      "start": {
        "column": 18,
        "index": 81,
        "line": 4,
      },
    },
    "local": "d",
    "source": "@vueuse/core",
    "subType": "id",
    "type": "import",
  },
]
`;

exports[`scan import > combine 1`] = `
{
  "exports": [],
  "imports": [
    {
      "loc": {
        "end": {
          "column": 16,
          "index": 17,
          "line": 2,
        },
        "start": {
          "column": 13,
          "index": 14,
          "line": 2,
        },
      },
      "local": "foo",
      "source": "bar",
      "type": "default",
    },
    {
      "loc": {
        "end": {
          "column": 22,
          "index": 51,
          "line": 3,
        },
        "start": {
          "column": 13,
          "index": 42,
          "line": 3,
        },
      },
      "local": "foo2",
      "source": "bar1",
      "type": "namespace",
    },
    {
      "imported": "foo3",
      "isType": false,
      "loc": {
        "end": {
          "column": 19,
          "index": 83,
          "line": 4,
        },
        "start": {
          "column": 15,
          "index": 79,
          "line": 4,
        },
      },
      "local": "foo3",
      "source": "bar2",
      "subType": "id",
      "type": "import",
    },
    {
      "imported": "foo3",
      "isType": false,
      "loc": {
        "end": {
          "column": 27,
          "index": 125,
          "line": 5,
        },
        "start": {
          "column": 15,
          "index": 113,
          "line": 5,
        },
      },
      "local": "foo4",
      "source": "bar3",
      "subType": "id",
      "type": "import",
    },
  ],
  "variables": [],
}
`;

exports[`scan import > import default 1`] = `
[
  {
    "loc": {
      "end": {
        "column": 10,
        "index": 10,
        "line": 1,
      },
      "start": {
        "column": 7,
        "index": 7,
        "line": 1,
      },
    },
    "local": "foo",
    "source": "bar",
    "type": "default",
  },
]
`;

exports[`scan import > import id 1`] = `
{
  "exports": [],
  "imports": [
    {
      "imported": "foo",
      "isType": false,
      "loc": {
        "end": {
          "column": 18,
          "index": 19,
          "line": 2,
        },
        "start": {
          "column": 15,
          "index": 16,
          "line": 2,
        },
      },
      "local": "foo",
      "source": "bar",
      "subType": "id",
      "type": "import",
    },
    {
      "imported": "foo1",
      "isType": false,
      "loc": {
        "end": {
          "column": 32,
          "index": 33,
          "line": 2,
        },
        "start": {
          "column": 20,
          "index": 21,
          "line": 2,
        },
      },
      "local": "foo2",
      "source": "bar",
      "subType": "id",
      "type": "import",
    },
    {
      "imported": "foo3",
      "isType": false,
      "loc": {
        "end": {
          "column": 48,
          "index": 49,
          "line": 2,
        },
        "start": {
          "column": 34,
          "index": 35,
          "line": 2,
        },
      },
      "local": "foo4",
      "source": "bar",
      "subType": "string",
      "type": "import",
    },
    {
      "imported": "bar1",
      "isType": false,
      "loc": {
        "end": {
          "column": 27,
          "index": 90,
          "line": 3,
        },
        "start": {
          "column": 15,
          "index": 78,
          "line": 3,
        },
      },
      "local": "bar2",
      "source": "bar",
      "subType": "id",
      "type": "import",
    },
    {
      "imported": "baz1",
      "isType": false,
      "loc": {
        "end": {
          "column": 29,
          "index": 133,
          "line": 4,
        },
        "start": {
          "column": 15,
          "index": 119,
          "line": 4,
        },
      },
      "local": "baz2",
      "source": "bar",
      "subType": "string",
      "type": "import",
    },
  ],
  "variables": [],
}
`;

exports[`scan import > import namespace 1`] = `
{
  "loc": {
    "end": {
      "column": 15,
      "index": 15,
      "line": 1,
    },
    "start": {
      "column": 7,
      "index": 7,
      "line": 1,
    },
  },
  "local": "foo",
  "source": "bar",
  "type": "namespace",
}
`;

exports[`scan import > import type 1`] = `
{
  "exports": [],
  "imports": [
    {
      "imported": "foo1",
      "isType": true,
      "loc": {
        "end": {
          "column": 24,
          "index": 25,
          "line": 2,
        },
        "start": {
          "column": 20,
          "index": 21,
          "line": 2,
        },
      },
      "local": "foo1",
      "source": "foo",
      "subType": "id",
      "type": "import",
    },
    {
      "imported": "foo2",
      "isType": true,
      "loc": {
        "end": {
          "column": 30,
          "index": 31,
          "line": 2,
        },
        "start": {
          "column": 26,
          "index": 27,
          "line": 2,
        },
      },
      "local": "foo2",
      "source": "foo",
      "subType": "id",
      "type": "import",
    },
    {
      "imported": "foo3",
      "isType": true,
      "loc": {
        "end": {
          "column": 24,
          "index": 69,
          "line": 3,
        },
        "start": {
          "column": 15,
          "index": 60,
          "line": 3,
        },
      },
      "local": "foo3",
      "source": "bar",
      "subType": "id",
      "type": "import",
    },
    {
      "imported": "foo4_real",
      "isType": false,
      "loc": {
        "end": {
          "column": 35,
          "index": 80,
          "line": 3,
        },
        "start": {
          "column": 26,
          "index": 71,
          "line": 3,
        },
      },
      "local": "foo4_real",
      "source": "bar",
      "subType": "id",
      "type": "import",
    },
  ],
  "variables": [],
}
`;
